scripts/rust-libc.py

```diff
@@ -0,0 +1,699 @@
+						if str(tc):
+							args.append(f"{Type.escape_name(arg_name)}: {str(tc)}")
+					if cursor.type.is_function_variadic():
+						args.append("...")
+					arg_str = ", ".join(args)
+					ret_type = str(Type(cursor, cursor.type.get_result()))
+					if not self.in_function_block:
+						emit("extern \"C\" {")
+						self.in_function_block = True
+					emit(f"\tpub fn {cursor.spelling}({arg_str})" + (f" -> {ret_type};" if ret_type != "::c_void" else ";"))
+					state.functions.append(cursor.spelling)
+				case CursorKind.TRANSLATION_UNIT:
+					for c in cursor.get_children():
+						self.from_cursor(header, c, filter_pred, level+1)
+				case CursorKind.INCLUSION_DIRECTIVE:
+					pass
+				case CursorKind.VAR_DECL:
+					if self.is_ignored("var_declarations", state.variables, cursor.spelling):
+						return
+					else:
+						log_err(f"munhandled cursor type", f"VAR_DECL of '{cursor.spelling}'")
+				case CursorKind.STATIC_ASSERT | CursorKind.UNEXPOSED_DECL:
+					pass
+				case CursorKind.MACRO_INSTANTIATION:
+					# TODO: cross-reference this with constant arrays?
+					pass
+				case _:
+					if not cursor.displayname.startswith("__"):
+						log_err("unhandled cursor type", f"{cursor.kind} {cursor.spelling} {cursor.displayname} {cursor.location}")
+
+						if t.is_valid():
+							emit(f"type '{t}' canonical '{t.canonical}'")
+		
+		if level == 0 and self.in_struct_block:
+			emit("}")
+			in_struct_block = False
+
+		if level == 0 and self.in_union_block:
+			emit("}")
+			in_union_block = False
+
+		if level == 0 and self.in_function_block:
+			emit("}")
+			in_function_block = False
+
+def parse(file: pathlib.Path, base_dir: pathlib.Path):
+	index = clang.cindex.Index.create()
+	tu = None
+
+	try:
+		tu = index.parse(base_dir / file,
+			args=[f"-I{p}" for p in config["includes"]] + ["-I" + str(base_dir)],
+			options=clang.cindex.TranslationUnit.PARSE_DETAILED_PROCESSING_RECORD | clang.cindex.TranslationUnit.PARSE_SKIP_FUNCTION_BODIES
+		)
+	except:
+		log_err("parsing error", file)
+		return
+
+	assert(tu)
+
+	if not RustBindingGenerator.is_ignored_file(base_dir, file, config):
+		if tu.diagnostics:
+			[log_err("compile error", d) for d in tu.diagnostics]
+			print(f"\n{errors_emitted + 1} errors emitted")
+			exit(errors_emitted + 1)
+
+		parser = RustBindingGenerator(config)
+		emit("")
+		print(f"// {tu.spelling.removeprefix(str(base_dir)).removeprefix("/")}")
+		parser.from_cursor(file, tu.cursor, no_system_includes)
+
+if __name__ == '__main__':
+	argparser = argparse.ArgumentParser()
+	argparser.add_argument("-n", dest="dry_run", action="store_true")
+	argparser.add_argument("path")
+	argparser.add_argument("file", nargs="?")
+
+	args = argparser.parse_args()
+
+	dry_run = args.dry_run
+
+	colorama.just_fix_windows_console()
+
+	with io.open("scripts/rust-libc-config.yml", "r") as f:
+		config = yaml.load(f, yaml.CSafeLoader)
+
+	path = pathlib.Path(args.path)
+
+	emit("// This file is autogenerated!")
```

Comment:
Perhaps this header could be read from a separate file, it's a bit easier to read that way + syntax highlighting etc?